manage_breast_screening/notifications/tests/queries/test_aggregate_query.py

```diff
@@ -0,0 +1,119 @@
+from datetime import datetime, timedelta
+from zoneinfo import ZoneInfo
+
+import pytest
+from django.db import connection
+
+from manage_breast_screening.notifications.models import Clinic
+from manage_breast_screening.notifications.queries.aggregate_query import AggregateQuery
+from manage_breast_screening.notifications.tests.factories import (
+    AppointmentFactory,
+    ChannelStatusFactory,
+    ClinicFactory,
+    MessageFactory,
+    MessageStatusFactory,
+)
+
+
+@pytest.mark.django_db
+class TestAggregateQuery:
+    def create_appointment_set(
+        self,
+        date: datetime,
+        clinic: Clinic,
+        episode_type: str,
+        message_status: str,
+        channel_statuses: dict[str, str],
+    ):
+        appt = AppointmentFactory(
+            clinic=clinic,
+            starts_at=date,
+            episode_type=episode_type,
+        )
+        message = MessageFactory(appointment=appt, status=message_status)
+        MessageStatusFactory(message=message, status=message_status)
+        for channel, status in channel_statuses.items():
+            ChannelStatusFactory(message=message, channel=channel, status=status)
+        return appt
+
+    def test_query_aggregates_appointments_and_cascade_counts(self):
+        clinic1 = ClinicFactory(code="BU000", bso_code="ABC", name="BSU 1")
+        clinic2 = ClinicFactory(code="BU001", bso_code="XYZ", name="BSU 2")
+        date1 = datetime.today() - timedelta(days=10)
+        date1.replace(tzinfo=ZoneInfo("Europe/London"))
+        date2 = datetime.today() - timedelta(days=15)
+        date2.replace(tzinfo=ZoneInfo("Europe/London"))
+
+        nhsapp_read = {"nhsapp": "read"}
+        sms_delivered = {"nhsapp": "unnotified", "sms": "delivered"}
+        letter_sent = {
+            "nhsapp": "unnotified",
+            "sms": "permanent_failure",
+            "letter": "received",
+        }
+        failed = {
+            "nhsapp": "unnotified",
+            "sms": "permanent_failure",
+            "letter": "validation_failed",
+        }
+
+        test_data = [
+            [date1, clinic1, "S", "delivered", nhsapp_read],
```

Comment:
small personal suggestion - i think I would find it easier to sense check the test if the test_data was organised in the rows expected, something like below would also help us ensure we've got all the scenarios covered.
```
            # row 1
            [date1, clinic1, "R", "failed", failed],
            # row 2
            [date1, clinic1, "S", "delivered", nhsapp_read],
            [date1, clinic1, "S", "delivered", sms_delivered],
            # row 3
            [date1, clinic2, "F", "delivered", nhsapp_read],
            [date1, clinic2, "F", "delivered", nhsapp_read],
            [date1, clinic2, "F", "delivered", sms_delivered],
            # row 4
            [date1, clinic2, "S", "delivered", nhsapp_read],
```

Reply:
@Harriethw  Have tried to make this clearer in https://github.com/NHSDigital/dtos-manage-breast-screening/pull/331/commits/82c31f962cd58557c3aa120fef37633e460d5886 and also count failures in the total.